value/eval.go

```diff
@@ -470,6 +470,143 @@ func ScanFirst(c Context, op string, v Value) Value {
 
 }
 
+// dataShape returns the data shape of v.
+// The data shape of a scalar is []int{}, to distinguish from a vector of length 1.
+func dataShape(v Value) []int {
+	switch v := v.(type) {
+	case Vector:
+		return []int{len(v)}
+	case *Matrix:
+		return v.Shape()
+	}
+	return []int{}
+}
+
+// appendData appends the underlying data of v to list.
+func appendData(list Vector, v Value) Vector {
+	switch v := v.(type) {
+	case Vector:
+		return append(list, v...)
+	case *Matrix:
+		return append(list, v.data...)
+	}
+	return append(list, v)
+}
+
+// BinaryMap computes the result of mapping op between lv and rv.
+func BinaryMap(c Context, lv Value, op string, rv Value) Value {
+	l := lv.toType(op, c.Config(), matrixType).(*Matrix)
+	r := rv.toType(op, c.Config(), matrixType).(*Matrix)
+	lsize := len(l.data)
+	rsize := len(r.data)
+	ln := 0
+	for ln < len(op) && op[ln] == '@' {
+		if l.shape[ln] > 0 {
+			lsize /= l.shape[ln]
+		}
+		ln++
+	}
+	rn := 0
+	for rn < len(op) && op[len(op)-1-rn] == '@' {
+		if r.shape[rn] > 0 {
+			rsize /= r.shape[rn]
+		}
+		rn++
+	}
+	var shape []int
+	var result Vector
+	for i := 0; i < len(l.data); i += lsize {
+		for j := 0; j < len(r.data); j += rsize {
+			var x Value = l.data[i : i+lsize]
+			var y Value = r.data[j : j+rsize]
+			if len(l.shape)-ln > 1 {
+				x = NewMatrix(l.shape[ln:], x.(Vector))
+			} else if len(l.shape)-ln == 0 {
+				x = x.(Vector).shrink()
+			}
+			if len(r.shape)-rn > 1 {
+				y = NewMatrix(r.shape[rn:], y.(Vector))
+			} else if len(r.shape)-rn == 0 {
+				y = y.(Vector).shrink()
+			}
+			a := c.EvalBinary(x, op[ln:len(op)-rn], y)
+			if shape == nil {
+				shape = dataShape(a)
+			} else {
+				if !sameShape(shape, dataShape(a)) {
+					Errorf("map %s: conflicting result shapes %s and %s",
+						op, NewIntVector(shape...), NewIntVector(dataShape(a)...))
+				}
+			}
+			result = appendData(result, a)
+		}
+	}
+	var mshape []int
+	mshape = append(mshape, l.shape[:ln]...)
+	mshape = append(mshape, r.shape[:rn]...)
+	if len(shape) > 1 || len(shape) == 1 && shape[0] > 1 {
+		mshape = append(mshape, shape...)
+	}
+	if len(mshape) == 1 {
+		return result
+	}
+	return NewMatrix(mshape, result)
+}
+
+// Map computes the result of mapping op onto v.
+// The trailing @ has been removed.
+func Map(c Context, op string, v Value) Value {
+	var shape []int
+	var result Vector
+	var n int
```

Comment:
rather C-like. i don't use 'var' much. also i usually compute in []Value and convert to vector at the end. if that makes sense here,  please convert it all. 

shape := []int{}
n := 0
data := []Value{}

Reply:
Interestingly, that's different from https://go.dev/s/style#declaring-empty-slices which suggests the var style for empty slices.

Reply:
Now there's just one. I left it as Vector because it is now always a Vector result.